File: November2021/C++/Unique_Binary_Search_Trees.cpp
//author : shyam2520
//https://leetcode.com/problems/unique-binary-search-trees/

class Solution {
public:
    int numTrees(int n) {
        vector<int> dict(n+1,0);
        dict[0]=1;
        dict[1]=1;
        // dict[2]=2;
        return findnode(n,dict);
    }
    
    int findnode(int n,vector<int>& dict)
    {
        if(dict[n])
            return dict[n];
        int ans=0;
        for(int i=1;i<=n;i++)
        {
            ans+=findnode(i-1,dict)*findnode(n-i,dict);
            
        }
        dict[n]=ans;
        return ans;
    }
};

File: November2021/UniqueBinarySearchTrees.java
class Solution {
	// TC : O(n2)
	private int[] dp = new int[20];
	public int numTrees(int n) {

		if(n <= 1) return 1;
		if(dp[n]>0) return dp[n];
		for(int i = 1; i <= n; i++)
			dp[n] += numTrees(i-1) * numTrees(n-i);
		return dp[n];

	}

}